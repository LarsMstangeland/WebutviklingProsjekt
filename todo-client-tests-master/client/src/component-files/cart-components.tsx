import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Card, Row, Column, Form, Button } from '../widgets';
import { createHashHistory } from 'history';
import cartService, {CartItem} from '../service-files/cart-service';
import { NavLink } from 'react-router-dom';
import recipeService from 'src/service-files/recipe-service';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student
//@ts-ignore
const userData = JSON.parse(sessionStorage.getItem('user'));

/**
 * Renders cart list.
 */
export class CartContent extends Component {
  //Array to store all cart
  cart:CartItem[] = [];
  CartItemsToShow:CartItem[] = [];
  searchBar: string = '';

  render() {
    return (
      <>
        <Card title="Your Cart">
          <Button.Danger onClick={() => {
            this.CartItemsToShow.map((cartitem) => {
              cartService.deleteIngredientFromCart(cartitem.cart_id)
            })
            this.mounted()
          }}>Clear All</Button.Danger>

          {this.CartItemsToShow.map((cart: CartItem) => (
            //Maps all the different cart and renders them as links to their respective cart details
            <Row key={cart.cart_id}>
              <Column>
                {cart.ingredients}
              </Column>
              <Column><Button.Danger onClick={() => {

                cartService.deleteIngredientFromCart(cart.cart_id).then(() => {
                  this.mounted();
                })
              }}>X</Button.Danger></Column>
            </Row>))}
        </Card>
      </>
    );
  }

  async mounted() {
    //Gets all cart and pass them to cart array
    try{
      let cart = await cartService.get(userData.user_id);
      this.cart = cart;
      this.CartItemsToShow = cart;

    } catch (error: any){
      Alert.danger('Error getting cart: ' + error.message)
    }
  }
}

