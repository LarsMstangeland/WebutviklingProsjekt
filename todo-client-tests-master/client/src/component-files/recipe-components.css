.react-simp {
  background-color: pink;
}

.picture-recipe {
  display: block;
  margin-left: auto;
  margin-right: auto;
  width: 50%;
}

/* The navbar container */
.topnav {
  overflow: hidden;
  background-color: #333;
}

/* Navbar links-------------mulig dette kan brukes */
.topnav a {
  float: left;
  display: block;
  color: #f2f2f2;
  text-align: center;
  padding: 14px 16px;
  text-decoration: none;
}

/* Links - change color on hover */
.topnav a:hover {
  background-color: #ddd;
  color: black;
}
