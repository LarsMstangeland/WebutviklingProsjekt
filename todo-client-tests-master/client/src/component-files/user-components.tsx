import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Card, Row, Column, Form, Button, PreviewCard } from '../widgets';
import { NavLink } from 'react-router-dom';
import userService, { User, LikedRecipe } from '../service-files/user-service';
import cartService, { CartItem } from '../service-files/cart-service';
import recipeService, {
  Ingredient,
  IngredientName,
  Recipe,
  RecipeToIngredient,
} from '../service-files/recipe-service';
import utilityService, { Region, Unit, Type } from '../service-files/utility-service';
import { createHashHistory } from 'history';
import bcrypt from 'bcryptjs';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student

//using this variable to check its value and then location.reload when true to make the user-info appear after creating new user
let created: boolean = false;

//@ts-ignore This is the userdata that gets added to sessionstorage if you log in. Ts-ignore because it can be empty
const userData = JSON.parse(sessionStorage.getItem('user'));

//function to hash password. To be done before adding password to database
export async function generateHash(password: string) {
  const salt = bcrypt.genSaltSync(10);
  const hash = bcrypt.hashSync(password, salt);
  return hash;
}

//function to compare entered password to the saved hashed password in the database
export async function compareHash(password: string, hashed: string) {
  return bcrypt.compareSync(password, hashed);
}

export class UserLogin extends Component {
  likedRecipes: LikedRecipe[] = [];
  users: User[] = [];
  loggedIn: boolean = false;
  user: User = { user_id: 0, username: '', password: '', admin: false };
  cart: CartItem[] = [];
  CartItemsToShow: CartItem[] = [];
  ingredients: Ingredient[] = [];
  newIngredient: IngredientName = { name: '', ingredients_id: 0 };
  newFridgeIngredient: IngredientName = { name: '', ingredients_id: 0 };
  searchBar: string = '';
  recipesToShow: Recipe[] = [];
  ingredientsToShow: IngredientName[] = [];
  fridgeIngredients: IngredientName[] = [];
  recipe: Recipe = {
    recipe_id: 0,
    name: '',
    description: '',
    region: '',
    picture_url: '',
    type: '',
  };
  recipes: Recipe[] = [];
  recipeIngredients: RecipeToIngredient[] = [];

  filterFridge() {
    let filteredRecipes: Recipe[] = [];
    for (let i = 0; i < this.fridgeIngredients.length; i++) {
      if (i == 0) {
        for (let j = 0; j < this.recipes.length; j++) {
          let recipeId = this.recipes[j].recipe_id;
          let ingredients = this.recipeIngredients.filter((row) => row.recipe_id == recipeId);
          for (let k = 0; k < ingredients.length; k++) {
            if (ingredients[k].ingredients_id == this.fridgeIngredients[i].ingredients_id) {
              filteredRecipes.push(this.recipes[j]);
            }
          }
        }
      } else {
        for (let j = 0; j < filteredRecipes.length; j++) {
          let recipeId = filteredRecipes[j].recipe_id;
          let ingredients = this.recipeIngredients.filter((row) => row.recipe_id == recipeId);
          if (
            ingredients.some(
              (ing) => ing.ingredients_id == this.fridgeIngredients[i].ingredients_id
            ) == false
          ) {
            filteredRecipes.splice(j, 1);
          }
        }
      }
    }
    filteredRecipes.map((recipe) => this.recipesToShow.push(recipe));
  }

  render() {
    // if userdata exists the page that renders is the one with your information
    if (userData) {
      return (
        <div style={{ display: 'flex', justifyContent: 'space-evenly', margin: '1vw' }}>
          <div style={{ display: 'flex', flexDirection: 'column', marginLeft: '1vw' }}>
            <div style={{ width: '45vw', marginTop: '2rem' }}>
              <Card title="Your user information">
                <Row>
                  <Column>Brukernavn: </Column>
                  <Column>{userData.username}</Column>
                </Row>
                <Row>
                  <Column>Access type: </Column>
                  <Column>{userData.admin ? 'Admin' : 'User'}</Column>
                </Row>
                <br />
                <Row>
                  <Column>Your liked recipes:</Column>
                </Row>
                {this.likedRecipes.length > 0
                  ? this.likedRecipes.map((recipe) => (
                      <Row key={recipe.recipe_id}>
                        <NavLink to={'/recipes/' + recipe.recipe_id}>
                          <Column>{recipe.name}</Column>
                        </NavLink>
                      </Row>
                    ))
                  : 'You have no liked recipes'}
                <br />
                <Row>
                  <Column>
                    <Button.Danger
                      onClick={() => {
                        sessionStorage.clear();
                        location.reload();
                      }}
                    >
                      Log out
                    </Button.Danger>
                  </Column>
                </Row>
              </Card>
            </div>
            {userData.admin ? (
              <div style={{ width: '45vw', marginTop: '2rem' }}>
                <Card title="Add ingredients">
                  <Row>
                    <Column>
                      <Form.Input
                        placeholder="Ingredient"
                        type="text"
                        value={this.newIngredient.name}
                        onChange={(event) => {
                          this.newIngredient.name = event.currentTarget.value;
                        }}
                      ></Form.Input>
                    </Column>
                    <Column>
                      <Button.Light
                        onClick={() => {
                          let name = this.newIngredient.name.toLowerCase();
                          if (this.ingredients.find((ing) => ing.name == name)) {
                            Alert.danger('This ingredient is already available');
                          } else {
                            recipeService.createIngredient(this.newIngredient);
                            this.newIngredient = { name: '', ingredients_id: 0 };
                            this.mounted();
                          }
                        }}
                      >
                        Add
                      </Button.Light>
                    </Column>
                    <Column>
                      <Button.Success
                        onClick={() => {
                          recipeService
                            .addRecipe(
                              this.recipe.name,
                              this.recipe.description,
                              this.recipe.picture_url,
                              this.recipe.region,
                              this.recipe.type
                            )
                            .then((response) => (this.recipe.recipe_id = response))
                            .then(() =>
                              history.push('/recipes/' + this.recipe.recipe_id + '/edit')
                            );
                        }}
                      >
                        Add Recipe
                      </Button.Success>
                    </Column>
                  </Row>
                </Card>
              </div>
            ) : (
              <></>
            )}
          </div>
          <div style={{ display: 'flex', flexDirection: 'column', marginLeft: '1vw' }}>
            <div style={{ marginTop: '2rem', width: '45vw' }}>
              <Card title="Your Cart">
                <br />
                {this.CartItemsToShow.map((cart: CartItem) => (
                  //Maps all the different cart and renders them as links to their respective cart details
                  <Row key={cart.cart_id}>
                    <Column>
                      {cart.ingredients.charAt(0).toUpperCase() + cart.ingredients.slice(1)}
                    </Column>
                    <Column>
                      <Button.Danger
                        small
                        onClick={() => {
                          cartService.deleteIngredientFromCart(cart.cart_id).then(() => {
                            this.mounted();
                          });
                        }}
                      >
                        X
                      </Button.Danger>
                    </Column>
                  </Row>
                ))}
                <br />
                <Button.Danger
                  onClick={() => {
                    this.CartItemsToShow.map((cartitem) => {
                      cartService.deleteIngredientFromCart(cartitem.cart_id);
                    });
                    this.mounted();
                  }}
                >
                  Clear All
                </Button.Danger>
              </Card>
            </div>
            <div style={{ width: '45vw', marginTop: '2rem' }}>
              <Card title="My Fridge">
                <>
                  <span>Find recipes based on the ingredients in your fridge!</span>
                  <Row>
                    <Column>
                      <Form.Input
                        type="search"
                        value={this.searchBar}
                        placeholder="Search for ingredient"
                        list={'searchList'}
                        onChange={(event) => {
                          this.searchBar = event.currentTarget.value;
                          //Filter function that finds the recipe which contain the same string as the searchbar, when there is more
                          //than one character in the searchbar
                          if (this.searchBar.length > 1) {
                            this.ingredientsToShow = [];
                            for (let i = 0; i < this.ingredients.length; i++) {
                              const name = this.ingredients[i].name.toUpperCase();
                              if (name.indexOf(this.searchBar.toUpperCase()) > -1) {
                                this.ingredientsToShow.push(this.ingredients[i]);
                              }
                            }
                          }
                        }}
                      ></Form.Input>
                      <datalist id="searchList">
                        {this.ingredientsToShow.map((ingredient) => (
                          //Maps all the ingredients that the filter function above selects and displays them as options in a datalist
                          <option key={ingredient.ingredients_id} value={ingredient.name}>
                            {ingredient.name}
                          </option>
                        ))}
                      </datalist>
                    </Column>
                    <Column>
                      <Button.Light
                        onClick={() => {
                          if (
                            this.ingredients.find((ing) => ing.name == this.searchBar.toLowerCase())
                          ) {
                            let ingredient: any = this.ingredients.find(
                              (ing) => ing.name == this.searchBar.toLowerCase()
                            );
                            if (
                              this.fridgeIngredients.find(
                                (ing) => ing.ingredients_id == ingredient?.ingredients_id
                              )
                            ) {
                              Alert.danger('This ingredient is already in your fridge');
                            } else {
                              this.fridgeIngredients.push(ingredient);
                              this.newFridgeIngredient = { name: '', ingredients_id: 0 };
                              this.recipesToShow = [];
                              this.filterFridge();
                            }
                          } else if (userData.admin) {
                            Alert.danger(
                              'This is not an ingredient, you can add ingredients in the left section of the page'
                            );
                          } else Alert.danger('This is not a ingredient');
                        }}
                      >
                        Add
                      </Button.Light>
                    </Column>
                  </Row>
                  {this.fridgeIngredients.map((ingredient: IngredientName) => (
                    //Maps all the different fridge ingredients and renders them as strings with button to remove
                    <Row key={ingredient.ingredients_id}>
                      <Column>
                        {ingredient.name.charAt(0).toUpperCase() + ingredient.name.slice(1)}
                      </Column>
                      <Column>
                        <Button.Danger
                          small
                          onClick={() => {
                            let index = this.fridgeIngredients.findIndex(
                              (ing) => ing.ingredients_id == ingredient.ingredients_id
                            );
                            this.fridgeIngredients.splice(index, 1);
                            this.recipesToShow = [];
                            this.filterFridge();
                          }}
                        >
                          X
                        </Button.Danger>
                      </Column>
                    </Row>
                  ))}

                  {this.recipesToShow.map((recipe) => (
                    <PreviewCard
                      small
                      key={recipe.recipe_id}
                      name={recipe.name}
                      url={recipe.picture_url}
                      id={recipe.recipe_id}
                    ></PreviewCard>
                  ))}
                </>
              </Card>
            </div>
          </div>
        </div>
      );
    }
    //if userdata does not exist, the page that renders is a login-page
    else {
      if (created == true) {
        created = false;
        location.reload();
      }
      return (
        <>
          <div
            style={{
              top: '30vh',
              width: '100%',
              display: 'flex',
              position: 'relative',
              alignItems: 'center',
              justifyContent: 'center',
              marginBottom: '1rem',
            }}
          >
            <Card title="Log in">
              <Row>
                <Column>
                  Brukernavn:
                  <Form.Input
                    type="text"
                    value={this.user.username}
                    onChange={(event) => {
                      this.user.username = event.currentTarget.value;
                    }}
                  ></Form.Input>
                </Column>
              </Row>
              <Row>
                <Column>
                  Passord:
                  <Form.Input
                    type="password"
                    value={this.user.password}
                    onChange={(event) => {
                      this.user.password = event.currentTarget.value;
                    }}
                  ></Form.Input>
                </Column>
              </Row>
              <Row>
                <Column>
                  <Button.Success
                    onClick={async () => {
                      if (this.users.find((u) => u.username == this.user.username)) {
                        let hashPass = this.users.find(
                          (u) => u.username == this.user.username
                        )?.password;
                        //@ts-ignore because hashpass can be undefined if the username typed in is not already a user
                        let hashCheck = await compareHash(this.user.password, hashPass);
                        if (hashCheck) {
                          //if the password is correct, pushes the info on a user to the sessionstorage.
                          //we decided to use the sessionstorage instead of creating unique apis for each user so that you cant type in the correct url and be taken to the page for a user, without actually logging in
                          let loggedInUser = this.users.find(
                            (u) => u.username == this.user.username
                          );
                          let userData = JSON.stringify(loggedInUser);
                          sessionStorage.setItem('user', userData);
                          //location.reload so that the page is drawn again, and the userinfo is portrayed instead of the login page                                        location.reload();
                          location.reload();
                        } else Alert.danger('Wrong username or password. Try again');
                      } else {
                        Alert.danger('Wrong username or password. Try again');
                      }
                    }}
                  >
                    Log in
                  </Button.Success>{' '}
                  <Button.Light onClick={() => history.push('/user/create')}>
                    Create user
                  </Button.Light>
                </Column>
              </Row>
            </Card>
          </div>
        </>
      );
    }
  }

  async mounted() {
    try {
      let users = await userService.getAll();
      this.users = users;
      let recipes = await recipeService.getAll();
      this.recipes = recipes;
      let recipeIngredients = await recipeService.getAllRecipeIngredients();
      //@ts-ignore
      this.recipeIngredients = recipeIngredients;
      let ingredients = await recipeService.getIngredients();
      //@ts-ignore
      this.ingredients = ingredients;

      if (userData) {
        let likedRecipes = await userService.getLikedRecipes(userData.user_id);
        this.likedRecipes = likedRecipes;

        try {
          let cart = await cartService.get(userData.user_id);
          this.cart = cart;
          this.CartItemsToShow = cart;
        } catch (error: any) {
          Alert.danger('Error getting cart: ' + error.message);
        }
      }
    } catch {
      Alert.danger('Could not fetch existing users from database');
    }
  }
}

export class NewUser extends Component {
  users: User[] = [];
  user: User = { user_id: 0, username: '', password: '', admin: false };
  passwordCheck: string = '';

  render() {
    return (
      <>
        <div
          style={{
            top: '30vh',
            width: '100%',
            display: 'flex',
            position: 'relative',
            alignItems: 'center',
            justifyContent: 'center',
            marginBottom: '1rem',
          }}
        >
          <Card title="Create new user">
            <Row>
                            <Column>
                                Username: 
                                <Form.Input 
                                // added id to find specific input for testing
                                id="username"
                                type="text" 
                                value={this.user.username} 
                                onChange={(event)=>{
                                    this.user.username = event.currentTarget.value;
                                }}></Form.Input>
                            </Column>
                        </Row>
                        <Row>
                            <Column>
                                Password:
                                <Form.Input
                                //using id in testing 
                                id="password"
                                type="password" 
                                value={this.user.password} 
                                onChange={(event)=>{
                                    this.user.password = event.currentTarget.value;
                                }}></Form.Input>
                            </Column>
                        </Row>
                        <Row>
                            <Column>
                                Confirm password:
                                <Form.Input 
                                //using id in testing 

                                id="passwordCheck"
                                type="password" 
                                value={this.passwordCheck} 
                                onChange={(event)=>{
                                    this.passwordCheck = event.currentTarget.value;
                                }}></Form.Input>
                            </Column>
                        </Row>
                        <Row>
                            <Column>
                            Admin: {' '}
                            <Form.Checkbox checked={this.user.admin} onChange={()=> {
                                this.user.admin == false ? this.user.admin = true : this.user.admin = false;
                            }
                            }></Form.Checkbox>
                            </Column>
                        </Row>
                        <Row>
                            <Column>
                            <Button.Success onClick={async ()=>{
                                if(!this.users.find(u => u.username == this.user.username)){
                                    if(this.user.password == this.passwordCheck){
                                        if(this.user.password != ''){
                                            if(this.user.username != ''){
                                                let hashPassword : string = await generateHash(this.user.password)
                                                await userService.create(hashPassword, this.user.username, this.user.admin) 
                                                 let u = await userService.get(this.user.username)
                                                 this.user = u       
                                                 sessionStorage.setItem('user', JSON.stringify(this.user));

                            // sets created to be true and then pushes to the UserLogin component. The UserLogin component will then be refreshed with the if-sentence on line 153
                            created = true;
                            history.push('/user/login');
                          } else {
                            Alert.danger("Username can't be null");
                          }
                        } else {
                          Alert.danger('Password cant be null');
                        }
                      } else {
                        Alert.danger("Passwords don't match");
                      }
                    } else {
                      Alert.danger('Username already exists. Try another one');
                    }
                  }}
                >
                  Create user
                </Button.Success>{' '}
                <Button.Danger onClick={() => history.push('/user/login')}>Cancel</Button.Danger>
              </Column>
            </Row>
          </Card>
        </div>
      </>
    );
  }

  async mounted() {
    try {
      let users = await userService.getAll();
      this.users = users;
    } catch {
      Alert.danger('Could not fetch existing users from database');
    }
  }
}
