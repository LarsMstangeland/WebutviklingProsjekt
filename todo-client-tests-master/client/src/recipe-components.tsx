import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Card, Row, Column, Form, Button } from './widgets';
import { NavLink } from 'react-router-dom';
import recipeService, { Recipe } from './recipe-service';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student

/**
 * Renders recipe list.
 */
export class RecipeList extends Component {
  recipes: Recipe[] = [];

  render() {
    return (
      <>
        <Card title="Recipes">
          {this.recipes.map((recipe) => (
            <Row key={recipe.id}>
              <Column>
                <NavLink to={'/recipes/' + recipe.id}>{recipe.name}</NavLink>
              </Column>
            </Row>
          ))}
        </Card>
      </>
    );
  }

  mounted() {
    recipeService
      .getAll()
      .then((recipes) => (this.recipes = recipes))
      .catch((error) => Alert.danger('Error getting recipes: ' + error.message));
  }
}

/**
 * Renders a specific recipe.
 */
export class RecipeDetails extends Component<{ match: { params: { id: number } } }> {
  recipe: Recipe = { id: 0, name: '', description: '', region: '' };

  render() {
    return (
      <>
        <Card title="recipe">
          <Row>
            <Column width={2}>Title:</Column>
            <Column>{this.recipe.name}</Column>
          </Row>
          <Row>
            <Column width={2}>Description:</Column>
          </Row>
          <Row>
            <Column width={2}>Region:</Column>
          </Row>
        </Card>
      </>
    );
  }

  mounted() {
    recipeService
      .get(this.props.match.params.id)
      .then((recipe) => (this.recipe = recipe))
      .catch((error) => Alert.danger('Error getting recipe: ' + error.message));
  }
}