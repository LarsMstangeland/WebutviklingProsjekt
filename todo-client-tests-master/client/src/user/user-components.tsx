import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type User = {
  id: number;
  name: string;
  cart: number;
};

class RecipeService {
  /**
   * Get recipe with given id.
   */
  get(id: number) {
    return axios.get<User>('/recipes/' + id).then((response) => response.data);
  }

  /**
   * Get all recipes.
   */
  getAll() {
    return axios.get<User[]>('/recipes').then((response) => response.data);
  }
}

const recipeService = new RecipeService();
export default recipeService;
