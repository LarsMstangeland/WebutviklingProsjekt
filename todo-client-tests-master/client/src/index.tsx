import ReactDOM from 'react-dom';
import * as React from 'react';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert, PreviewCard } from './widgets';
import { RecipeList, RecipeDetails, RecipeEdit } from './component-files/recipe-components';
import {UserLogin} from './component-files/user-components';

class Menu extends Component {
  render() {
    return (
      <NavBar brand="Food Junkies">
        <NavBar.Link to="/recipes">Recipes</NavBar.Link>
        <NavBar.Link to="/my_profile">My Profile</NavBar.Link>
      </NavBar>
    );
  }
}

class Home extends Component {
  render() {
    return (
      <div>
        <Card title="Welcome">To Food Junkies</Card>

          <PreviewCard title1 ="dette er en tittel1" title2 = "dette er en tittel2" image1 = "https://i.guim.co.uk/img/media/26392d05302e02f7bf4eb143bb84c8097d09144b/446_167_3683_2210/master/3683.jpg?width=1200&quality=85&auto=format&fit=max&s=a52bbe202f57ac0f5ff7f47166906403" image2 = "https://i.guim.co.uk/img/media/26392d05302e02f7bf4eb143bb84c8097d09144b/446_167_3683_2210/master/3683.jpg?width=1200&quality=85&auto=format&fit=max&s=a52bbe202f57ac0f5ff7f47166906403">
          </PreviewCard>


      </div>

    )

  }
}

ReactDOM.render(
  <HashRouter>
    <div>
      <Alert />
      <Menu />
      <Route exact path="/" component={Home} />
      <Route exact path="/recipes" component={RecipeList} />
      <Route exact path="/recipes/:id(\d+)" component={RecipeDetails} /> {/* id must be number */}
      <Route exact path='/recipes/:id(\d+)/edit' component={RecipeEdit} />
      <Route exact path='/my_profile' component={UserLogin}/>
    </div>
  </HashRouter>,
  document.getElementById('root')
);
