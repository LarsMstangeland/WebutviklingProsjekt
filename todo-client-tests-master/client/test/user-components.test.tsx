import * as React from 'react';
import { shallow } from 'enzyme';
import { NavLink } from 'react-router-dom';
import {UserLogin, NewUser} from '../src/component-files/user-components';
import userService, {User, LikedRecipe} from '../src/service-files/user-service';
import { Form , Button, NavBar} from '../src/widgets';

jest.mock('../src/service-files/user-service', () => {
    class UserService {
        get() {
            return Promise.resolve(
                {user_id: 1, username: 'vetle', password: 'tevle', admin: true}
            );
        }

        getAll() {
            return Promise.resolve([
                {user_id: 1, username : 'vetle', password : 'tevle', admin : true},
                {user_id : 2, username : 'sebastian', password : 'seb', admin : false}
            ]);
        }

        create() {
            return Promise.resolve(3);
        }

        delete() {
            return Promise.resolve();
        }

    }
})

// jest.setTimeout(30000);

describe('Components draw correctly tests' , () => {
    test('UserLogin draws correctly' , () => {
        const wrapper = shallow(<UserLogin/>);

        expect(wrapper).toMatchSnapshot();
    });

    // test('NewUser draws correctly' , () => {
    //     const wrapper = shallow(<NewUser/>);

    //         expect(wrapper).toMatchSnapshot();
    // });
});

describe('Testing NewUser-component', () => {
    test('Input fields value changes', (done) => {
        const wrapper = shallow(<NewUser></NewUser>);

        setTimeout(() => {
            expect(wrapper.containsAllMatchingElements([
                //@ts-ignore
                <Form.Input value=""></Form.Input>,
                //@ts-ignore
                <Form.Input value=""></Form.Input>,
                //@ts-ignore
                <Form.Input value=""></Form.Input>,
            ])).toEqual(true);
        });

        //@ts-ignore
        wrapper.find('#username').simulate('change', {currentTarget : {value : 'vetleek'}});                
        //@ts-ignore
        expect(wrapper.containsMatchingElement(<Form.Input value="vetleek"></Form.Input>)).toEqual(true);

        //@ts-ignore
        wrapper.find('#password').simulate('change', {currentTarget : {value : 'pass'}});                
        //@ts-ignore
        expect(wrapper.containsMatchingElement(<Form.Input value="pass"></Form.Input>)).toEqual(true);

        //@ts-ignore
        wrapper.find('#passwordCheck').simulate('change', {currentTarget : {value : 'pass'}});                
        //@ts-ignore
        expect(wrapper.containsMatchingElement(<Form.Input value="pass"></Form.Input>)).toEqual(true);

        done();
    });

    test('Checkbox changes value of admin onclick', (done) => {
        const wrapper = shallow(<NewUser></NewUser>);

        setTimeout(() => {
            expect(wrapper.containsMatchingElement(
                //@ts-ignore
                <Form.Checkbox checked={false}></Form.Checkbox>
            ));
        });

        wrapper.find(Form.Checkbox).simulate('change',  {checked : true});
        //@ts-ignore
        expect(wrapper.containsMatchingElement(<Form.Checkbox checked={true}></Form.Checkbox>));
        done();
    });

    test('Button correctly sets location on click', (done) => {
        const wrapper = shallow(<NewUser></NewUser>);

        wrapper.find(Button.Success).simulate('click');
        

        setTimeout(() => {
            expect(location.hash).toEqual('#/user/login');
        });
        done();
    });
});
