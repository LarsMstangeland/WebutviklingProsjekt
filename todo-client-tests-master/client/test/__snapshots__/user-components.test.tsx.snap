// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components draw correctly tests NewUser draws correctly 1`] = `
<Fragment>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
        "marginBottom": "1rem",
        "position": "relative",
        "top": "30vh",
        "width": "100%",
      }
    }
  >
    <Card
      title="Create new user"
    >
      <Row>
        <Column>
          Username:
          <FormInput
            onChange={[Function]}
            type="text"
            value=""
          />
        </Column>
      </Row>
      <Row>
        <Column>
          Password:
          <FormInput
            onChange={[Function]}
            type="password"
            value=""
          />
        </Column>
      </Row>
      <Row>
        <Column>
          Confirm password:
          <FormInput
            onChange={[Function]}
            type="password"
            value=""
          />
        </Column>
      </Row>
      <Row>
        <Column>
          Admin: 
           
          <FormCheckbox
            checked={false}
            onChange={[Function]}
          />
        </Column>
      </Row>
      <Row>
        <Column>
          <ButtonSuccess
            onClick={[Function]}
          >
            Create user
          </ButtonSuccess>
           
          <ButtonDanger
            onClick={[Function]}
          >
            Cancel
          </ButtonDanger>
        </Column>
      </Row>
    </Card>
  </div>
</Fragment>
`;

exports[`Components draw correctly tests UserLogin draws correctly when logged in 1`] = `
<div
  style={
    {
      "display": "flex",
      "justifyContent": "space-evenly",
      "margin": "1vw",
    }
  }
>
  <div
    style={
      {
        "display": "flex",
        "flexDirection": "column",
        "marginLeft": "1vw",
      }
    }
  >
    <div
      style={
        {
          "marginTop": "2rem",
          "width": "45vw",
        }
      }
    >
      <Card
        title="Your user information"
      >
        <Row>
          <Column>
            Brukernavn: 
          </Column>
          <Column />
        </Row>
        <Row>
          <Column>
            Access type: 
          </Column>
          <Column>
            Admin
          </Column>
        </Row>
        <br />
        <Row>
          <Column>
            Your liked recipes:
          </Column>
        </Row>
        You have no liked recipes
        <br />
        <Row>
          <Column>
            <ButtonDanger
              onClick={[Function]}
            >
              Log out
            </ButtonDanger>
          </Column>
        </Row>
      </Card>
    </div>
    <div
      style={
        {
          "marginTop": "2rem",
          "width": "45vw",
        }
      }
    >
      <Card
        title="Add ingredients"
      >
        <Row>
          <Column>
            <FormInput
              onChange={[Function]}
              placeholder="Ingredient"
              type="text"
              value=""
            />
          </Column>
          <Column>
            <ButtonLight
              onClick={[Function]}
            >
              Add
            </ButtonLight>
          </Column>
          <Column>
            <ButtonSuccess
              onClick={[Function]}
            >
              Add Recipe
            </ButtonSuccess>
          </Column>
        </Row>
      </Card>
    </div>
  </div>
  <div
    style={
      {
        "display": "flex",
        "flexDirection": "column",
        "marginLeft": "1vw",
      }
    }
  >
    <div
      style={
        {
          "marginTop": "2rem",
          "width": "45vw",
        }
      }
    >
      <Card
        title="Your Cart"
      >
        <br />
        <br />
        <ButtonDanger
          onClick={[Function]}
        >
          Clear All
        </ButtonDanger>
      </Card>
    </div>
    <div
      style={
        {
          "marginTop": "2rem",
          "width": "45vw",
        }
      }
    >
      <Card
        title="My Fridge"
      >
        <span>
          Find recipes based on the ingredients in your fridge!
        </span>
        <Row>
          <Column>
            <FormInput
              list="searchList"
              onChange={[Function]}
              placeholder="Search for ingredient"
              type="search"
              value=""
            />
            <datalist
              id="searchList"
            />
          </Column>
          <Column>
            <ButtonLight
              onClick={[Function]}
            >
              Add
            </ButtonLight>
          </Column>
        </Row>
      </Card>
    </div>
  </div>
</div>
`;

exports[`Components draw correctly tests UserLogin draws correctly when not logged in 1`] = `
<Fragment>
  <div
    style={
      {
        "alignItems": "center",
        "display": "flex",
        "justifyContent": "center",
        "marginBottom": "1rem",
        "position": "relative",
        "top": "30vh",
        "width": "100%",
      }
    }
  >
    <Card
      title="Log in"
    >
      <Row>
        <Column>
          Brukernavn:
          <FormInput
            onChange={[Function]}
            type="text"
            value=""
          />
        </Column>
      </Row>
      <Row>
        <Column>
          Passord:
          <FormInput
            onChange={[Function]}
            type="password"
            value=""
          />
        </Column>
      </Row>
      <Row>
        <Column>
          <ButtonSuccess
            onClick={[Function]}
          >
            Log in
          </ButtonSuccess>
           
          <ButtonLight
            onClick={[Function]}
          >
            Create user
          </ButtonLight>
        </Column>
      </Row>
    </Card>
  </div>
</Fragment>
`;
